### PR TITLE
pybind: add verbose error message

### PR DESCRIPTION
**Pull request by iotcg**
Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/pybind/rbd/rbd.pyx [hunk 1]
```diff
@@ -700,71 +700,110 @@ class OSError(Error):
         return (self.__class__, (self.message, self.errno))
 
 class PermissionError(OSError):
-    pass
+    def __init__(self, msg):
```
Comment (dillaman) [inline src/pybind/rbd/rbd.pyx]: You dropped the ```errno``` keyword argument, so all of these will fail to be constructed in ```make_ex```

**Comment by tchaikov** [comment]
@iotcg ping?

**Comment by iotcg** [comment]
@dillaman Could you give more info about what's wrong with the patches?

**Comment by iotcg** [comment]
@tchaikov I'm waiting for dillaman's reply.

**Comment by dillaman** [comment]
@iotcg Just see my comment -- it will crash and burn when an exception is thrown.

File: src/pybind/rbd/rbd.pyx (hunk 1 above)
Comment (tchaikov) [reply src/pybind/rbd/rbd.pyx]: @iotcg i think what Jason means is that before your change, the exception classes inherited from `OSError` will can constructed using `make_ex()` which is used everywhere in this module. and this function does use the `errno` keyword argument. your change simply removes the inherited constructor. and it almost breaks all the callers of `make_ex()`.

File: src/pybind/rados/rados.pyx [hunk 2]
```diff
@@ -894,7 +943,7 @@ Rados object in state %s." % self.state)
         # for now and remove it later
         with nogil:
             ret = rados_connect(self.cluster)
-        if ret != 0:
+        if (ret != 0 and ret != -errno.EISCONN):
```
Comment (tchaikov) [inline src/pybind/rados/rados.pyx]: i'd suggest remove the parenthesis around condition.

Comment (tchaikov) [reply src/pybind/rados/rados.pyx]: please note, this changes the behavior of this method. shall we consider it a bug in the application which calls `Rados.connect()` with an already-connected rados instance?

File: src/pybind/rados/rados.pyx [hunk 3]
```diff
@@ -351,80 +351,125 @@ class Error(Exception):
         return (self.__class__, (self.message, self.errno))
 
 class InvalidArgumentError(Error):
-    pass
+    def __init__(self, message, errno=None):
+        super(InvalidArgumentError, self).__init__(
+            ("RADOS Invalid Argument error (%s)" % message), errno)
```
Comment (tchaikov) [inline src/pybind/rados/rados.pyx]: i'd suggest remove the "error" in the error message. as it's always an error in this context.

File: src/init-ceph.in [hunk 4]
```diff
@@ -447,6 +447,10 @@ for name in $what; do
 	    [ -n "$asok" ] && rm -f $asok
 	    [ -n "$post_stop" ] && do_cmd "$post_stop"
 	    [ -n "$lockfile" ] && [ "$?" -eq 0 ] && rm -f $lockfile
+	    # flush all data out of journal in background
```
Comment (tchaikov) [inline src/init-ceph.in]: could you elaborate the rationale behind this change in the commit message? also, might want to fix the typo in the title of the commit message.

File: src/ceph_common.sh [hunk 5]
```diff
@@ -182,10 +182,14 @@ get_name_list() {
...
-	$CCONF -c $conf -l mgr | egrep -v '^mgr$' || true ; \
-	$CCONF -c $conf -l osd | egrep -v '^osd$' || true`
+    tmp=$(mktemp /tmp/ceph.XXXXXXX)
+    echo $local >> $tmp
+    $CCONF -c $conf -l mon | egrep -v '^mon$' >> $tmp || true
+    $CCONF -c $conf -l mds | egrep -v '^mds$' >> $tmp || true
+    $CCONF -c $conf -l mgr | egrep -v '^mgr$' >> $tmp || true
+    $CCONF -c $conf -l osd | egrep -v '^osd$' >> $tmp || true
+    allconf=`cat $tmp | xargs -n 1 | sort -u`
+    rm $tmp
```
Comment (tchaikov) [inline src/ceph_common.sh]: nit, could put
```sh
allconf=$(for entity in \
  $local \
  `$CCONF -c $conf -l mon | egrep -v '^mon$' || true` \
  `$CCONF -c $conf -l mds | egrep -v '^mds$' || true` \
  `$CCONF -c $conf -l mgr | egrep -v '^mgr$' || true` \
  `$CCONF -c $conf -l osd | egrep -v '^osd$' || true`; do
  echo $entity
done | sort -u)
```

File: src/pybind/rados/rados.pyx (hunk 3 above)
Comment (iotcg) [reply src/pybind/rados/rados.pyx]: @tchaikov I'll change it

File: src/pybind/rados/rados.pyx (hunk 2 above)
Comment (iotcg) [reply src/pybind/rados/rados.pyx]: 1. I'll remove the parenthesis
2. I'll check more detail about this method

File: src/init-ceph.in (hunk 4 above)
Comment (iotcg) [reply src/init-ceph.in]: I'll refind the commit message

Comment (iotcg) [reply src/init-ceph.in]: @tchaikov Previously, there's some unknown problem in ceph-v10.2.6 and this patch is used. I'm not sure what the problem is(long time ago) and whether the problem will be still hit in ceph master branch. Is it proper to still upstream this patch?

File: src/ceph_common.sh (hunk 5 above)
Comment (iotcg) [reply src/ceph_common.sh]: thanks for suggestion

File: src/pybind/rados/rados.pyx (hunk 2 above)
Comment (iotcg) [reply src/pybind/rados/rados.pyx]: No. I'll remove the change.

File: src/init-ceph.in (hunk 4 above)
Comment (tchaikov) [reply src/init-ceph.in]: @iotcg it depends on if we consider it's an improvement. if you are not sure if we have this problem and we cannot reproduce it anymore. i don't see any reason to fix it.

Comment (iotcg) [reply src/init-ceph.in]: Agree. Unclear change is not necessary.

**Comment by iotcg** [comment]
@dillaman ping

**Comment by tchaikov** [comment]
retest this please.

File: src/pybind/rbd/rbd.pyx [hunk 6]
```diff
@@ -700,71 +700,111 @@ class OSError(Error):
...
+    def __init__(self, message, errno=None):
+        super(Timeout, self).__init__(
+                ("RBD operation timeout (%s)" % message), errno)
+
 
 class DiskQuotaExceeded(OSError):
-    pass
+    def __init__(self, message, errno=None):
+        super(DiskQuotaExceeded, self).__init__(
+                ("RBD Disk Quota Exceeded (%s)" % message), errno)
```
Comment (dillaman) [inline src/pybind/rbd/rbd.pyx]: Nit: `RBD disk quota exceeded ...` (not CamelCase)

Comment (iotcg) [reply src/pybind/rbd/rbd.pyx]: Solved. Thanks

**Comment by iotcg** [comment]
retest this please

File: src/pybind/rbd/rbd.pyx [hunk 7]
```diff
@@ -700,71 +700,111 @@ class OSError(Error):
...
+    def __init__(self, message, errno=None):
+        super(Timeout, self).__init__(
+                ("RBD operation timeout (%s)" % message), errno)
+
 
 class DiskQuotaExceeded(OSError):
-    pass
+    def __init__(self, message, errno=None):
+        super(DiskQuotaExceeded, self).__init__(
+                ("RBD disk quota exceeded (%s)" % message), errno)
```
Comment (tchaikov) [inline src/pybind/rbd/rbd.pyx]: @iotcg just a nit before running this change through qa suites: could you remove the parentheses around the message. in the case of `DiskQuotaExceeded`, i'd suggest remove the parentheses around "RBD disk quota exceeded (%s)" % message".

Comment (iotcg) [reply src/pybind/rbd/rbd.pyx]: @tchaikov  Which one do you prefer? A or B
A. ("RBD disk quota exceeded (%s)" % message) ==> "RBD disk quota exceeded %s" % message
B. ("RBD disk quota exceeded (%s)" % message) ==> "RBD disk quota exceeded (%s)" % message

Comment (tchaikov) [reply src/pybind/rbd/rbd.pyx]: @iotcg the latter.

Comment (iotcg) [reply src/pybind/rbd/rbd.pyx]: Thanks. I'll follow this suggestion in ceph.

**Comment by iotcg** [comment]
any feedback?

**Comment by tchaikov** [comment]
@iotcg sorry for the latency. i am adding your PR to my batch.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2019-07-22_16:29:12-rados-wip-kefu-testing-2019-07-22-1948-distro-basic-mira/

http://pulpito.ceph.com/kchai-2019-07-24_01:28:15-rados-wip-kefu-testing-2019-07-22-1948-distro-basic-smithi/

- "evicting unresponsive client": tracked by http://tracker.ceph.com/issues/40867
- tasks.mgr.test_progress.TestProgress: addressed by #29111
- tasks.mgr.dashboard.test_health.HealthTest: already fixed by #29172

**Pull request merged**